fw/vhost.c

```diff
@@ -2572,6 +2644,170 @@ tfw_vhost_cfgclean(void)
+                                 capo->arg);
+                        break;
+                case TFW_D_CACHE_FULFILL:
+                        T_LOG_NL("      cache_fulfill %s\n",
+                                 capo->arg);
+                        break;
+                case TFW_D_CACHE_RESP_HDR_DEL:
+                        T_LOG_NL("      cache_resp_hdr_del %s\n",
+                                 capo->arg);
+                        break;
+                case TFW_D_CACHE_CONTROL_IGNORE:
+                        T_LOG_NL("      cache_control_ignore %s\n",
+                                 capo->arg);
+                        break;
+                }
+
+        }
+
+        for (i = 0; i < loc->nipdef_sz; ++i) {
+                TfwNipDef *nipdef= loc->nipdef[i];
+                T_LOG_NL("      %s\n", nipdef->arg);
+        }
+
+        if (h_mods) {
+                for (i = 0; i < h_mods->sz; ++i) {
+                        TfwHdrModsDesc *d = &h_mods->hdrs[i];
+                        T_LOG_NL("      %s  %i  %i\n", d->hdr->data,
+                                d->hdr->nchunks, (int)d->hdr->len);
+                }
+        }
+        T_LOG_NL("   }");
+}
+
+static void
+tfw_cfgop_frang_global_print(FrangGlobCfg *frang_gconf)
+{
+	if (!frang_gconf)
+		return;
+
+	T_LOG_NL("frang_limits {\n");
+	T_LOG_NL("   clnt_hdr_timeout=%lu\n",
+		 frang_gconf->clnt_hdr_timeout);
+	T_LOG_NL("   clnt_body_timeout=%lu\n",
+		 frang_gconf->clnt_body_timeout);
+	T_LOG_NL("   req_rate=%u\n", frang_gconf->req_rate);
+	T_LOG_NL("   req_burst=%u\n", frang_gconf->req_burst);
+	T_LOG_NL("   conn_rate=%u\n", frang_gconf->conn_rate);
+	T_LOG_NL("   conn_burst_rate=%u\n", frang_gconf->conn_burst);
+	T_LOG_NL("   conn_max=%u\n", frang_gconf->conn_max);
+	T_LOG_NL("   tls_new_conn_rate=%u\n",
+		 frang_gconf->tls_new_conn_rate);
+	T_LOG_NL("   tls_new_conn_burst=%u\n",
+		 frang_gconf->tls_new_conn_burst);
+	T_LOG_NL("   tls_incomplete_conn_rate=%u\n",
+		 frang_gconf->tls_incomplete_conn_rate);
+	T_LOG_NL("   http_hchunk_cnt=%u\n",
+		 frang_gconf->http_hchunk_cnt);
+	T_LOG_NL("   http_bchunk_cnt=%u\n",
+		 frang_gconf->http_bchunk_cnt);
+	T_LOG_NL("   http_hdr_len=%u\n",
+		 frang_gconf->http_hdr_len);
+	T_LOG_NL("   http_hdr_cnt=%u\n",
+		 frang_gconf->http_hdr_cnt);
+	T_LOG_NL("   ip_block=%s\n",
+		 frang_gconf->ip_block ? "true" : "false");
+	T_LOG_NL("}\n");
+}
+
+static void
+tfw_cfgop_vhosts_print(TfwVhostList *vhosts)
+{
+	TfwVhost *vhost;
+	char str[128];
+	int i, j;
+	int len;
+
+	T_LOG_NL("Actual configuration.\n");
+	if (!vhosts)
+		return;
+
+	memset(str, 0, sizeof(str));
+	tfw_cfgop_frang_global_print(&tfw_frang_glob_reconfig);
+
+	T_LOG_NL("tls_match_any_server_name=%s\n",
+		 tfw_tls_get_allow_any_sni_reconfig() ? "true" : "false");
+	hash_for_each(vhosts->vh_hash, i, vhost, hlist) {
+                len = vhost->name.len < 128 ? vhost->name.len : 127;
+                memcpy(str, vhost->name.data, len);
```

Comment:
Why we can't use `strncpy` here and do not call  memset(str, 0, len); every time?

Reply:
I agree memset is redundant here, but, seems, memcpy and strncpy are very similar in this case.